### PR TITLE
Avoid allocating on case-insensitive string comparisons.

### PR DESCRIPTION
**Pull request by DanielRosenwasser**

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

File: src/compiler/core.ts [hunk 1]
```diff
@@ -1773,9 +1773,27 @@ namespace ts {
         if (a === b) return Comparison.EqualTo;
         if (a === undefined) return Comparison.LessThan;
         if (b === undefined) return Comparison.GreaterThan;
-        a = a.toUpperCase();
-        b = b.toUpperCase();
-        return a < b ? Comparison.LessThan : a > b ? Comparison.GreaterThan : Comparison.EqualTo;
+        const aLength = a.length;
+        const bLength = b.length;
+        const maxLength = Math.min(aLength, bLength);
+        for (let i = 0; i < maxLength; i++) {
```
Comment (dmichon-msft) [inline src/compiler/core.ts:1779]: One of the perks of operating this way is that you could extend the function to take in the start and end indices of each string as arguments, such that places where you do a substring comparison can avoid the substring allocation. Not sure how much it matters in practice, though.

**Comment by dmichon-msft** [comment]
Using an `Intl.Collator` object to do the comparison is about 10% slower than the reference in microbenchmarks, but should theoretically avoid the allocations and also offers additional options for sort behavior.

File: src/compiler/core.ts [hunk 2]
```diff
@@ -1773,9 +1773,27 @@ namespace ts {
         if (a === b) return Comparison.EqualTo;
         if (a === undefined) return Comparison.LessThan;
         if (b === undefined) return Comparison.GreaterThan;
-        a = a.toUpperCase();
-        b = b.toUpperCase();
-        return a < b ? Comparison.LessThan : a > b ? Comparison.GreaterThan : Comparison.EqualTo;
+        const aLength = a.length;
+        const bLength = b.length;
+        const maxLength = Math.min(aLength, bLength);
```
Comment (amcasey) [inline src/compiler/core.ts:1778]: "minLength"?

**Review by amcasey — COMMENTED** [review]
I would be interested to know whether, when you compile something like the compiler itself, "cheap" cases outnumber "expensive" cases.  i.e. This slows down the slow case, so it would be reassuring to know that the slow case is uncommon (in English code).

File: src/compiler/core.ts (hunk 2 above)
Comment (DanielRosenwasser) [reply src/compiler/core.ts:1778]: Capping values always has this unintuitive aspect. The max index of two strings is always determined by the min of the two lengths. Similarly, if you needed to cap a potentially-negative index to `0`, you'd call it a `minIndex`, but call `Math.max`.

Comment (amcasey) [reply src/compiler/core.ts:1778]: "maxIndex" would be fine, but it's the smaller (i.e. minimum) of the two lengths.

Comment (DanielRosenwasser) [reply src/compiler/core.ts:1778]: Well...it's not a valid index 😄

`lengthUpperBound`?

Comment (dmichon-msft) [reply src/compiler/core.ts:1778]: `commonLength`?

**Pull request closed without merging**